Content/LeagueSandbox-Scripts/Champions/Taric/E.cs

```diff
@@ -1,73 +1,92 @@
+            // TODO
+        };
+
+        public void OnActivate(IObjAiBase owner, ISpell spell)
         {
         }
 
-        public void OnDeactivate(IObjAiBase owner)
+        public void OnDeactivate(IObjAiBase owner, ISpell spell)
         {
         }
 
-        public void OnStartCasting(IObjAiBase owner, ISpell spell, IAttackableUnit target)
+        public void OnSpellPreCast(IObjAiBase owner, ISpell spell, IAttackableUnit target, Vector2 start, Vector2 end)
         {
+        }
 
+        public void OnSpellCast(ISpell spell)
+        {
         }
 
-        public void OnFinishCasting(IObjAiBase owner, ISpell spell, IAttackableUnit target)
+        public void OnSpellPostCast(ISpell spell)
         {
-            if (Vector2.DistanceSquared(owner.Position, target.Position) > 625 * 625)
+            if (Vector2.DistanceSquared(spell.CastInfo.Owner.Position, spell.CastInfo.Targets[0].Unit.Position) > 625 * 625)
             {
                 return;
             }
             else
             {
-                spell.AddProjectileTarget("Dazzle", target, true);
+                //spell.AddProjectileTarget("Dazzle", spell.CastInfo.SpellCastLaunchPosition, spell.CastInfo.Targets[0].Unit, HitResult.HIT_Normal, true);
             }
         }
 
-        public void ApplyEffects(IObjAiBase owner, IAttackableUnit target, ISpell spell, ISpellMissile projectile)
+        public void ApplyEffects(IObjAiBase owner, IAttackableUnit target, ISpell spell, ISpellMissile missile)
         {
-            var time = 1.1f + 0.1f * spell.Level;
+            var time = 1.1f + 0.1f * spell.CastInfo.SpellLevel;
             var ap = owner.Stats.AbilityPower.Total;
-            var damage = 10 + spell.Level * 30 + ap * 0.2f;
+            var damage = 10 + spell.CastInfo.SpellLevel * 30 + ap * 0.2f;
             var dist = Vector2.DistanceSquared(owner.Position, target.Position);
             if (dist <= 460 * 460)
             {
-                damage = 15 + spell.Level * 45 + ap * 0.3f;
+                damage = 15 + spell.CastInfo.SpellLevel * 45 + ap * 0.3f;
             }
             if (dist <= 295 * 295)
             {
-                damage = 20 + spell.Level * 60 + ap * 0.4f;
+                damage = 20 + spell.CastInfo.SpellLevel * 60 + ap * 0.4f;
             }
 
             target.TakeDamage(owner, damage, DamageType.DAMAGE_TYPE_MAGICAL, DamageSource.DAMAGE_SOURCE_SPELL, false);
 
-            AddBuff("TaricEHud", time, 1, spell, (IObjAiBase)target, owner);
-            AddBuff("Stun", time, 1, spell, (IObjAiBase)target, owner);
+            AddBuff("TaricEHud", time, 1, spell, target, owner);
+            AddBuff("Stun", time, 1, spell, target, owner);
             AddParticleTarget(owner, "Dazzle_tar.troy", target);
-            var p102 = AddParticleTarget(owner, "Global_Stun.troy", target, 1.25f, "head");
             var p103 = AddParticleTarget(owner, "Taric_HammerFlare.troy", target, 1);
 
             CreateTimer(time, () =>
```

Comment:
The particles remove themselves automatically or not?

Reply:
They do. Most scripts have not been updated, so things like CreateTimer, or manual removals of particles have not been removed.